File: server/src/models/agent.ts
import { z } from 'zod';

// Agent 사고 모드
export const ThinkingModeSchema = z.enum([
  'idle',         // 대기 - 작업 없음
  'exploring',    // 탐색 - 정보 수집 중
  'structuring',  // 구조화 - 작업 분해/정리 중
  'validating',   // 검증 - 결과 확인 중
  'summarizing',  // 요약 - 결과 정리 중
]);

export type ThinkingMode = z.infer<typeof ThinkingModeSchema>;

// Agent 유형
export const AgentTypeSchema = z.enum([
  'document-processor',   // 문서 처리
  'email-handler',        // 이메일 처리
  'research-assistant',   // 리서치
  'schedule-manager',     // 일정 관리
  'task-coordinator',     // 작업 조율
  'custom',               // 커스텀
]);

export type AgentType = z.infer<typeof AgentTypeSchema>;

// Agent 상태
export const AgentStatusSchema = z.enum([
  'active',       // 활성 - 작업 수행 중
  'idle',         // 대기 - 작업 없음
  'paused',       // 일시정지 - 사용자 입력 대기
  'error',        // 오류 상태
  'disabled',     // 비활성화
]);

export type AgentStatus = z.infer<typeof AgentStatusSchema>;

// Agent 제약조건
export const AgentConstraintSchema = z.object({
  id: z.string().uuid(),
  type: z.enum([
    'action_forbidden',     // 금지된 행동
    'approval_required',    // 승인 필수
    'notify_required',      // 알림 필수
    'limit_scope',          // 범위 제한
    'time_restriction',     // 시간 제한
  ]),
  description: z.string(),
  condition: z.string().optional(),  // 조건식 (선택적)
  isActive: z.boolean().default(true),
  source: z.enum(['ontology', 'user', 'system']).default('system'),
});

export type AgentConstraint = z.infer<typeof AgentConstraintSchema>;

// Agent 스키마
export const AgentSchema = z.object({
  id: z.string().uuid(),
  name: z.string().min(1),
  type: AgentTypeSchema,
  description: z.string().optional(),

  // 상태 정보
  status: AgentStatusSchema.default('idle'),
  thinkingMode: ThinkingModeSchema.default('idle'),

  // 현재 작업
  currentTaskId: z.string().uuid().optional(),
  currentTaskDescription: z.string().optional(),

  // 제약조건
  constraints: z.array(AgentConstraintSchema).default([]),

  // 권한
  permissions: z.object({
    canCreateTickets: z.boolean().default(true),
    canExecuteApproved: z.boolean().default(true),
    canAccessMcp: z.array(z.string()).default([]),  // 접근 가능한 MCP 서비스
  }).default({}),

  // 통계
  stats: z.object({
    ticketsCreated: z.number().default(0),
    ticketsCompleted: z.number().default(0),
    ticketsRejected: z.number().default(0),
    averageApprovalTime: z.number().optional(),  // ms
  }).default({}),

  // 메타데이터
  lastActivity: z.date(),
  createdAt: z.date(),
  updatedAt: z.date(),
});

export type Agent = z.infer<typeof AgentSchema>;

// Agent 등록 입력
export const RegisterAgentInputSchema = AgentSchema.omit({
  id: true,
  status: true,
  thinkingMode: true,
  currentTaskId: true,
  currentTaskDescription: true,
  stats: true,
  lastActivity: true,
  createdAt: true,
  updatedAt: true,
}).partial({
  constraints: true,
  permissions: true,
});

export type RegisterAgentInput = z.infer<typeof RegisterAgentInputSchema>;

// Agent 상태 업데이트
export const AgentStateUpdateSchema = z.object({
  agentId: z.string().uuid(),
  status: AgentStatusSchema.optional(),
  thinkingMode: ThinkingModeSchema.optional(),
  currentTaskId: z.string().uuid().optional().nullable(),
  currentTaskDescription: z.string().optional().nullable(),
});

export type AgentStateUpdate = z.infer<typeof AgentStateUpdateSchema>;

File: server_python/models/agent.py
from enum import Enum
from datetime import datetime
from typing import Optional, List
from uuid import UUID, uuid4
from pydantic import BaseModel, Field


class ThinkingMode(str, Enum):
    IDLE = "idle"
    EXPLORING = "exploring"
    STRUCTURING = "structuring"
    VALIDATING = "validating"
    SUMMARIZING = "summarizing"


class AgentType(str, Enum):
    DOCUMENT_PROCESSOR = "document-processor"
    EMAIL_HANDLER = "email-handler"
    RESEARCH_ASSISTANT = "research-assistant"
    SCHEDULE_MANAGER = "schedule-manager"
    TASK_COORDINATOR = "task-coordinator"
    CUSTOM = "custom"


class AgentStatus(str, Enum):
    ACTIVE = "active"
    IDLE = "idle"
    PAUSED = "paused"
    ERROR = "error"
    DISABLED = "disabled"


class ConstraintType(str, Enum):
    ACTION_FORBIDDEN = "action_forbidden"
    APPROVAL_REQUIRED = "approval_required"
    NOTIFY_REQUIRED = "notify_required"
    LIMIT_SCOPE = "limit_scope"
    TIME_RESTRICTION = "time_restriction"


class ConstraintSource(str, Enum):
    ONTOLOGY = "ontology"
    USER = "user"
    SYSTEM = "system"


class AgentConstraint(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    type: ConstraintType
    description: str
    condition: Optional[str] = None
    isActive: bool = True
    source: ConstraintSource = ConstraintSource.SYSTEM


class AgentPermissions(BaseModel):
    canCreateTickets: bool = True
    canExecuteApproved: bool = True
    canAccessMcp: List[str] = Field(default_factory=list)


class AgentStats(BaseModel):
    ticketsCreated: int = 0
    ticketsCompleted: int = 0
    ticketsRejected: int = 0
    averageApprovalTime: Optional[float] = None  # ms


class Agent(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    name: str
    type: AgentType
    description: Optional[str] = None
    
    # 상태 정보
    status: AgentStatus = AgentStatus.IDLE
    thinkingMode: ThinkingMode = ThinkingMode.IDLE
    
    # 현재 작업
    currentTaskId: Optional[str] = None
    currentTaskDescription: Optional[str] = None
    
    # 제약조건
    constraints: List[AgentConstraint] = Field(default_factory=list)
    
    # 권한
    permissions: AgentPermissions = Field(default_factory=AgentPermissions)
    
    # 통계
    stats: AgentStats = Field(default_factory=AgentStats)
    
    # 메타데이터
    lastActivity: datetime = Field(default_factory=datetime.now)
    createdAt: datetime = Field(default_factory=datetime.now)
    updatedAt: datetime = Field(default_factory=datetime.now)


class RegisterAgentInput(BaseModel):
    name: str
    type: AgentType
    description: Optional[str] = None
    constraints: Optional[List[dict]] = None
    permissions: Optional[dict] = None


class AgentStateUpdate(BaseModel):
    agentId: str
    status: Optional[AgentStatus] = None
    thinkingMode: Optional[ThinkingMode] = None
    currentTaskId: Optional[str] = None
    currentTaskDescription: Optional[str] = None

